presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/IntermediateAggregationOptimizer.java

```diff
@@ -0,0 +1,310 @@
+
+            // Re-introduce the aggregation as an INTERMEDIATE step
+            // No need to translate specification since above exchange will output same output symbols as the original
+            AggregationSpecification aggregation = context.get().get();
+            rewrittenPlanNode = aggregation.applyAfter(idAllocator, INTERMEDIATE, rewrittenPlanNode);
+
+            // Re-apply the current REMOTE GATHER exchange
+            // The Optional<AggregationSpecification> should have been normalized to have the same output symbols as its inputs.
+            // This means the input and output symbols to this exchange will be the same.
+            rewrittenPlanNode = new ExchangeNode(
+                    node.getId(),
+                    node.getType(),
+                    node.getScope(),
+                    node.getPartitioningScheme(),
+                    ImmutableList.of(rewrittenPlanNode),
+                    ImmutableList.of(node.getOutputSymbols()));
+
+            return new Result<>(rewrittenPlanNode, Boolean.FALSE);
+        }
+
+        private Function<Symbol, Symbol> outputToInputTranslator(ExchangeNode node, int sourceIndex)
+        {
+            return symbol -> node.getInputs().get(sourceIndex).get(node.getOutputSymbols().indexOf(symbol));
+        }
+
+        @Override
+        public Result<Boolean> visitTableScan(TableScanNode node, RewriteContext<Optional<AggregationSpecification>, Boolean> context)
+        {
+            return new Result<>(node, Boolean.TRUE);
+        }
+
+        @Override
+        public Result<Boolean> visitUnion(UnionNode node, RewriteContext<Optional<AggregationSpecification>, Boolean> context)
+        {
+            return context.defaultRewrite(node, Optional.empty(), this::retainIfAny);
+        }
+
+        @Override
+        public Result<Boolean> visitJoin(JoinNode node, RewriteContext<Optional<AggregationSpecification>, Boolean> context)
+        {
+            return context.defaultRewrite(node, Optional.empty(), this::retainFirstSource);
+        }
+
+        @Override
+        public Result<Boolean> visitSemiJoin(SemiJoinNode node, RewriteContext<Optional<AggregationSpecification>, Boolean> context)
+        {
+            return context.defaultRewrite(node, Optional.empty(), this::retainFirstSource);
+        }
+
+        @Override
+        public Result<Boolean> visitIndexJoin(IndexJoinNode node, RewriteContext<Optional<AggregationSpecification>, Boolean> context)
+        {
+            return context.defaultRewrite(node, Optional.empty(), this::retainFirstSource);
+        }
+
+        private Boolean retainFirstSource(List<Boolean> booleans)
+        {
+            return booleans.get(0);
+        }
+
+        private Boolean retainIfAny(List<Boolean> booleans)
+        {
+            return booleans.stream().anyMatch(isSourcePartitioned -> isSourcePartitioned);
+        }
+    }
+
+    private static class AggregationSpecification
```

Comment:
I think I'm inclined to hold off on this until we see that we need this in at least one more use case. But I'm not too strongly opinionated about this.